File: C++Verifier/src/expressions/AddExpr.cpp
#ifndef ADDEXPR_H
#define ADDEXPR_H

#include <vector>
#include "ArithExpr.cpp"

using namespace std;

class AddExpr: public ArithExpr {
    public:
        using ArithExpr::ArithExpr;

        void* evaluate(shared_ptr<Packet> p) {
            string leftValue = evalSingle(left, leftType, p);
            string rightValue = evalSingle(right, rightType, p);
            string *ret = nullptr;
            *ret = applyArithOp(leftValue, rightValue, "+");
            return static_cast<void*>(ret);
        }

        string evaluate(shared_ptr<Packet> p, vector<string> constraints, vector<string> locationList, vector<string> variableList) {
            string leftValue = evalSingle(left, leftType, p, constraints, locationList, variableList);
            string rightValue = evalSingle(right, rightType, p, constraints, locationList, variableList);
            return applyArithOp(leftValue, rightValue, "+");
        }
        string toString() {
            return "(" + (left -> toString()) + " + " + (right -> toString()) + ")";
        }
};

#endif

File: verification/src/main/java/org/msr/mnr/verification/expressions/AddExpr.java
package org.msr.mnr.verification.expressions;

import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;

import org.msr.mnr.verification.dsfa.ConstraintTreeNode;
import org.msr.mnr.verification.utils.Packet;

public class AddExpr extends ArithExpr {
    private static final long serialVersionUID = 4L;

    public AddExpr(Object left, Object right) {
        super(left, right);
    }

    @Override
    public int[] evaluate(Packet p) {
        int[] leftValue = evalSingle(left, leftType, p);
        int[] rightValue = evalSingle(right, rightType, p);
        return applyArithOp(leftValue, rightValue, "+");
    }


    @Override
    public int[] evaluate(Packet p, ArrayList<int[]> constraints,
            ArrayList<String> locationList, ArrayList<String> variableList) {

        int[] leftValue = evalSingle(left, leftType, p, constraints, locationList, variableList);
        int[] rightValue = evalSingle(right, rightType, p, constraints, locationList, variableList);
        return applyArithOp(leftValue, rightValue, "+");
    }

    @Override
    public String toString() {
        return "(" + left.toString() + " + " + right.toString() + ")";
    }
}
